api/commands/test.ts

```diff
@@ -0,0 +1,77 @@
+import { Credentials } from "@dataform/api/commands/credentials";
+import * as dbadapters from "@dataform/api/dbadapters";
+import { dataform } from "@dataform/protos";
+
+export async function test(
+  graph: dataform.ICompiledGraph,
+  credentials: Credentials
+): Promise<dataform.ITestResult[]> {
+  const dbadapter = dbadapters.create(credentials, graph.projectConfig.warehouse);
+  return await Promise.all(graph.tests.map(testCase => runTest(dbadapter, testCase)));
+}
+
+async function runTest(
+  dbadapter: dbadapters.DbAdapter,
+  testCase: dataform.ITest
+): Promise<dataform.ITestResult> {
+  const [actualResults, expectedResults] = await Promise.all([
+    dbadapter.execute(testCase.testQuery),
+    dbadapter.execute(testCase.expectedOutputQuery)
+  ]);
+
+  // Check row counts.
+  if (actualResults.length !== expectedResults.length) {
+    return {
+      name: testCase.name,
+      successful: false,
+      message: `Expected ${expectedResults.length} rows, but saw ${actualResults.length} rows.`
+    };
+  }
+  if (actualResults.length === 0) {
+    return {
+      name: testCase.name,
+      successful: true
+    };
+  }
+
+  // Check column sets.
+  const actualColumns = Object.keys(actualResults[0]);
+  const expectedColumns = Object.keys(expectedResults[0]);
+  if (actualColumns.length !== expectedColumns.length) {
+    return {
+      name: testCase.name,
+      successful: false,
+      message: `Expected columns "${expectedColumns}", but saw "${actualColumns}".`
+    };
+  }
+  for (const expectedColumn of expectedColumns) {
```

Comment:
OK so this is fun - different warehouses care differently about case sensitivity. In fact, snowflake ALWAYS returns results in uppercase, as you probably noticed.

I'm pretty sure bigquery is case insensitive, let's check redshift. If it is too, then we should probably try and deal with this.